search/agent/consumer.py

```diff
@@ -0,0 +1,408 @@
+import os
+from typing import List, Any, Optional
+from search import logging
+from search.services import metadata, index
+from search.process import transform
+from search.domain import DocMeta, Document, asdict
+from .base import BaseRecordProcessor, ProcessRecordsInput
+
+logger = logging.getLogger(__name__)
+
+
+class DocumentFailed(RuntimeError):
+    """Raised when an arXiv paper could not be added to the search index."""
+
+
+class IndexingFailed(RuntimeError):
+    """Raised when indexing failed such that future success is unlikely."""
+
+
+class MetadataRecordProcessor(BaseRecordProcessor):
+    """Consumes ``MetadataIsAvailable`` notifications, updates the index."""
+
+    MAX_ERRORS = 5
+    """Max number of individual document failures before aborting entirely."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialize exception counter."""
+        super(MetadataRecordProcessor, self).__init__(*args, **kwargs)  # type: ignore
+        self._error_count = 0
+        self._cache: Optional[str] = None
+
+    def init_cache(self, cache_dir: str) -> None:
+        """Configure the processor to use a local cache for docmeta."""
+        if not os.path.exists(cache_dir):
+            raise ValueError(f'cache_dir does not exist: {cache_dir}')
+        self._cache = cache_dir
+
+    def _from_cache(self, arxiv_id: str) -> DocMeta:
+        """
+        Get the docmeta document from a local cache, if available.
+
+        Parameters
+        ----------
+        arxiv_id : str
+
+        Returns
+        -------
+        :class:`.DocMeta`
+
+        Raises
+        ------
+        RuntimeError
+            Raised when the cache is not available, or the document could not
+            be found in the cache.
+
+        """
+        if not self._cache:
+            raise RuntimeError('Cache not set')
+
+        fname = '%s.json' % arxiv_id.replace('/', '_')
+        cache_path = os.path.join(self._cache, fname)
+        if not os.path.exists(cache_path):
+            raise RuntimeError('No cached document')
+
+        with open(cache_path) as f:
+            data: dict = json.load(f)
+            return DocMeta(**data)  # type: ignore
+            # See https://github.com/python/mypy/issues/3937
+
+    def _to_cache(self, arxiv_id: str, docmeta: DocMeta) -> None:
+        """
+        Add a document to the local cache, if available.
+
+        Parameters
+        ----------
+        arxiv_id : str
+        docmeta : :class:`.DocMeta`
+
+        Raises
+        ------
+        RuntimeError
+            Raised when the cache is not available, or the document could not
+            be added to the cache.
+
+        """
+        if not self._cache:
+            raise RuntimeError('Cache not set')
+        fname = '%s.json' % arxiv_id.replace('/', '_')
+        cache_path = os.path.join(self._cache, fname)
+        try:
+            with open(cache_path, 'w') as f:
+                json.dump(asdict(docmeta).items(), f)
```

Comment:
I don't think .items() is needed here ("Object of type 'dict_items' is not JSON serializable").

Reply:
Doh. Indeed